org.eclipse.gef.examples.ui.pde/src/org/eclipse/gef/examples/ui/pde/internal/wizards/ProjectUnzipperNewWizard.java

```diff
@@ -208,7 +202,7 @@ public void execute(IProgressMonitor monitor) throws InterruptedException {
 								 * Copy plug-in project code
 								 */
 								extractProject(projectFolderFile, getProjectZipURL()[i],
-										new SubProgressMonitor(monitor, 100));
+										SubMonitor.convert(monitor, 100));
```

Comment:
This is a really common pattern I've seen countless times by now, which sadly breaks the IProgressMonitor contract. From the JavaDoc:

```
Since it is illegal to call beginTask on the same IProgressMonitor more than once, the same instance of IProgressMonitor must not be passed to convert more than once.
```

Instead I would go with something like this:
```java

public boolean performFinish() {

    try {
        IRunnableWithProgress operation = new WorkspaceModifyOperation() {

            @Override
            public void execute(IProgressMonitor monitor) throws InterruptedException {
                try {
                    SubMonitor subMonitor = SubMonitor.convert(monitor, Messages.monitor_creatingProject, 150);

                    /*
                        * Create the project folder
                        */
                    IPath projectPath = wizardNewProjectCreationPage.getLocationPath();

                    for (int i = 0; i < nameFormats.length; i++) {
                        String projectName = MessageFormat.format(nameFormats[i],
                                wizardNewProjectCreationPage.getProjectName());
                        String projectFolder = projectPath.toOSString() + File.separator + projectName;
                        File projectFolderFile = new File(projectFolder);

                        IWorkspace workspace = ResourcesPlugin.getWorkspace();
                        IProject project = workspace.getRoot().getProject(projectName);

                        // If the project does not exist, we will create it
                        // and populate it.
                        if (!project.exists()) {
                            projectFolderFile.mkdirs();
                            subMonitor.worked(10);

                            /*
                                * Copy plug-in project code
                                */
                            extractProject(projectFolderFile, getProjectZipURL()[i], subMonitor.newChild(100));

                            if (subMonitor.isCanceled()) {
                                throw new InterruptedException();
                            }

                            if (projectPath.equals(workspace.getRoot().getLocation())) {
                                project.create(subMonitor.newChild(10));
                            } else {
                                IProjectDescription desc = workspace.newProjectDescription(project.getName());
                                desc.setLocation(new Path(projectFolder));

                                project.create(desc, subMonitor.newChild(10));
                            }
                        }

                        // Now, we ensure that the project is open.
                        project.open(subMonitor.newChild(10));

                        renameProject(project, projectName);

                        // Add Java and PDE natures
                        IProjectDescription desc = workspace.newProjectDescription(project.getName());
                        desc.setNatureIds(
                                new String[] { ORG_ECLIPSE_PDE_PLUGIN_NATURE, ORG_ECLIPSE_JDT_CORE_JAVANATURE });
                        project.setDescription(desc, subMonitor.newChild(10));

                        subMonitor.worked(10);
                        if (subMonitor.isCanceled()) {
                            throw new InterruptedException();
                        }
                    } // end for

                } catch (IOException e) {
                    throw new RuntimeException(e);
                } catch (CoreException e) {
                    throw new RuntimeException(e);
                }
            }
        };

        getContainer().run(false, true, operation);

        /* Set perspective */
        BasicNewProjectResourceWizard.updatePerspective(getConfigurationElement());

    } catch (InterruptedException e) {
        return false;

    } catch (Exception e) {

        GefExamplesPlugin.getDefault().getLog().log(new Status(IStatus.ERROR, "org.eclipse.gef.examples.ui.pde", //$NON-NLS-1$
                IStatus.OK, e.getMessage(), e));
        return false;
    }

    return true;
}
```

It's also a bad idea to pass the same monitor to e.g. `project.open()`, because it may internally create a SubMonitor (or even just call `done()`). Meaning you also have to create a sub-monitor for each of those calls.

Reply:
Thx for pointing that us. This is important that I learned that. Is the last comment also the reason you removed the monitor.done in the finally?

Reply:
> This is important that I learned that. Is the last comment also the reason you removed the monitor.done in the finally?

Partially, yes. Calling `done()` is generally the responsibility of whoever created the IProgressMonitor instance. So this is something that should be handled by the caller, not the callee.

The SubMonitor API specifically says that you don't need to call `done()` on the created SubMonitor. And there is also no need, given that all of the allocated ticks are consumed by the time the method finishes.

Reply:
Thx for the clarification. This is very helpful.

Reply:
forgot to mention. Changed everything as you suggested and did some more clean-ups. Also tested it in a dedicated Eclipse instance.

Reply:
To be honest, most of my knowledge comes from the [Using Progress Monitor](https://www.eclipse.org/articles/Article-Progress-Monitors/article.html) article, as well as trial and error 😅 

I think this is the interesting part:
# 4.0 Responsibilities of callers and callees
Imagine you need to invoke another method that accepts an IProgressMonitor. What are the responsibilities of the caller and what are the responsibilities of the callee?

The caller:

- Will pass an IProgressMonitor instance which has not had beginTask invoked on it yet (or an implementation such as SubMonitor which permits beginTask to be invoked multiple times).
- Will not expect that the callee invokes done() on the monitor. The caller must either use an SubMonitor (or a similar implementation which does not require done() to be invoked), or it must take responsibility for calling done() on the monitor after the callee has finished.
- Will not pass in a null monitor unless the JavaDoc for the callee says that it accepts null.
- Will pass in a monitor which ignores the String argument to beginTask unless the JavaDoc for the callee says that it requires otherwise.

The callee:

- Will call beginTask 0 or 1 times on the monitor, at its option.
- Will not promise to call done() on the monitor, although it is allowed to do so.
- Will not call setCanceled on the monitor.
- Will not accept a null monitor unless its JavaDoc says otherwise.
- Will not expect the caller to do anything with the string passed to beginTask unless its JavaDoc says otherwise.

Reply:
Guess I should read that in more detail.